File: c#/FuturesButterfly/FuturesButterfly/PositionManager.cs
using MySql.Data.MySqlClient;
using StrategyUtilities;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FuturesButterfly
{
    public static class PositionManager
    {
        public static PositionManagerOutput GetFuturesButterflyPosition(string alias, DateTime asOfDate, MySqlConnection conn)
        {
            DataTable StrategyPosition = TA.Strategy.GetNetPosition(alias: alias, conn: conn, asOfDate: asOfDate);
            StrategyUtilities.PositionManagerOutput PositionManagerOut = new StrategyUtilities.PositionManagerOutput();

            if (StrategyPosition.Rows.Count == 0)
            {
                return PositionManagerOut;
            }

            StrategyPosition.Columns.Add("TickerHead", typeof(string));
            StrategyPosition.Columns.Add("SortIndex", typeof(int));

            foreach (DataRow Row in StrategyPosition.Rows)
            {
                ContractUtilities.ContractSpecs Cs = ContractUtilities.ContractMetaInfo.GetContractSpecs(ticker: Row.Field<string>("Ticker"));
                Row["TickerHead"] = Cs.tickerHead;
                Row["SortIndex"] = Cs.contINDX;
            }

            StrategyPosition = DataAnalysis.DataTableFunctions.Sort(dataTableInput: StrategyPosition, columnList: new string[] { "SortIndex" });
            bool CorrectPositionQ = false;

            if ((StrategyPosition.Rows.Count == 3) &&
                (Math.Sign(StrategyPosition.Rows[0].Field<double>("Qty"))) == Math.Sign(StrategyPosition.Rows[2].Field<double>("Qty")) &&
                (StrategyPosition.AsEnumerable().Sum(x => x.Field<double>("Qty")) == 0))
            {
                CorrectPositionQ = true;
            }

            PositionManagerOut.SortedPosition = StrategyPosition;
            PositionManagerOut.Scale = StrategyPosition.Rows[0].Field<double>("Qty");
            PositionManagerOut.CorrectPositionQ = CorrectPositionQ;
            return PositionManagerOut;
        }
    }
}

File: PycharmProjects/ta/position_manager.py

import ta.strategy as ts
import contract_utilities.contract_meta_info as cmi

def get_ocs_position(**kwargs):

    net_position = ts.get_net_position_4strategy_alias(**kwargs)

    empty_position_q = False
    correct_position_q = False
    scale = 0

    if len(net_position.index) == 0:
        empty_position_q = True
        return {'empty_position_q':empty_position_q, 'correct_position_q': correct_position_q, 'scale': scale,'sorted_position': net_position}

    net_position['cont_indx'] = [cmi.get_contract_specs(x)['cont_indx'] for x in net_position['ticker']]
    net_position.sort_values('cont_indx', ascending=True, inplace=True)

    if (len(net_position.index) == 2) & (net_position['qty'].sum() == 0):
        correct_position_q = True
        scale = net_position['qty'].iloc[0]

    return {'empty_position_q': empty_position_q, 'correct_position_q': correct_position_q, 'scale': scale,'sorted_position': net_position}

